src/appengine/libs/handler.py

```diff
@@ -415,7 +415,113 @@ def wrapper(self, *args, **kwargs):
       raise helpers.AccessDeniedException('Invalid CSRF token.')
 
     # Make sure that the token is not expired.
-    elif token.expiration_time < datetime.datetime.utcnow():
+    if token.expiration_time < datetime.datetime.utcnow():
+      token.key.delete()
+      raise helpers.AccessDeniedException('Expired CSRF token.')
+
+    return func(self, *args, **kwargs)
+
+  return wrapper
+
+
+def extend_request_flask(req, params):
+  """Extends a request."""
+
+  def _iterparams():
+    for k, v in six.iteritems(params):
+      yield k, v
+
+  req.iterparams = _iterparams
+
+
+def extend_json_request_flask(req):
+  """Extends a request to support JSON."""
+  try:
+    params = json.loads(req.data)
+    req.args = params
+  except ValueError:
+    raise helpers.EarlyExitException(
+        'Parsing the JSON request body failed: %s' % req.body, 400)
+
+  extend_request_flask(req, req.args)
+
+
+def post_flask(request_content_type, response_content_type):
```

Comment:
Let's put all these into a separate module so we can just call these post() etc. When the time comes to complete migration, all we need to do is rename the module.

Reply:
done